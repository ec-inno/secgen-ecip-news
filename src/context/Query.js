import React, { useContext } from 'react';

import { defaultLangKey } from '../../languages';

const queryInit = {
  filters: {},
  status: 'ALL',
  language: defaultLangKey,
  pagination: '0/10',
};

const queryReducer = (state, action) => {
  switch (action.type) {
    case 'paginate': {
      return {
        ...state,
        pagination: action.pagination,
      };
    }

    case 'changeLanguage': {
      return {
        ...state,
        pagination: queryInit.pagination,
        language: action.language,
      };
    }

    case 'setFilters': {
      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...state.filters,
          ...action.filters,
        },
      };
    }

    case 'unsetFilter': {
      const { filters } = state;
      delete filters[action.filter];

      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...filters,
        },
      };
    }

    case 'reset':
    default: {
      return queryInit;
    }
  }
};

const Context = React.createContext();

const useQueryContext = () => useContext(Context);

export { queryInit, queryReducer, Context as default, useQueryContext };
