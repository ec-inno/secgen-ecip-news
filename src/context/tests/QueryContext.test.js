import { queryInit, queryReducer } from '../Query';

describe('QueryContext', () => {
  it('exports initial state', () => {
    expect(queryInit).toMatchSnapshot();
  });

  it('filters can be set through the reducer', () => {
    const state = queryReducer(queryInit, {
      type: 'setFilters',
      filters: { foo: 'bar' },
    });
    expect(state).toMatchSnapshot();
  });

  it('filter can be removed through the reducer', () => {
    const stateFiltersAdded = queryReducer(queryInit, {
      type: 'setFilters',
      filters: { foo: 'bar', baz: 'bar' },
    });

    expect(stateFiltersAdded.filters.foo).toBe('bar');
    expect(stateFiltersAdded.filters.baz).toBe('bar');

    const stateFiltersRemoved = queryReducer(stateFiltersAdded, {
      type: 'unsetFilter',
      filter: 'foo',
    });

    expect(stateFiltersRemoved.filters.foo).toBe(undefined);
    expect(stateFiltersRemoved.filters.baz).toBe('bar');
  });

  it('section can be changed through the reducer', () => {
    const state = queryReducer(queryInit, {
      type: 'changeSection',
      section: 'LATEST',
    });
    expect(state).toMatchSnapshot();
  });

  it('language can be changed through the reducer', () => {
    const state = queryReducer(queryInit, {
      type: 'changeLanguage',
      language: 'bg',
    });
    expect(state).toMatchSnapshot();
  });

  it('pagination is possible through the reducer', () => {
    const state = queryReducer(queryInit, {
      type: 'paginate',
      pagination: '11/20',
    });
    expect(state).toMatchSnapshot();
  });

  it('state can be reset through the reducer', () => {
    const stateFiltersAdded = queryReducer(queryInit, {
      type: 'setFilters',
      filters: { foo: 'bar', baz: 'bar' },
    });

    expect(stateFiltersAdded.filters.foo).toBe('bar');
    expect(stateFiltersAdded.filters.baz).toBe('bar');

    const stateReset = queryReducer(stateFiltersAdded, {
      type: 'reset',
    });

    expect(stateReset).toEqual(queryInit);
  });
});
