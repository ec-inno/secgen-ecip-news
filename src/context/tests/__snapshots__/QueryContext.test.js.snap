// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryContext exports initial state 1`] = `
Object {
  "filters": Object {},
  "language": "en",
  "pagination": "0/10",
  "section": "ALL",
}
`;

exports[`QueryContext filters can be set through the reducer 1`] = `
Object {
  "filters": Object {
    "foo": "bar",
  },
  "language": "en",
  "pagination": "0/10",
  "section": "ALL",
}
`;

exports[`QueryContext language can be changed through the reducer 1`] = `
Object {
  "filters": Object {},
  "language": "bg",
  "pagination": "0/10",
  "section": "ALL",
}
`;

exports[`QueryContext pagination is possible through the reducer 1`] = `
Object {
  "filters": Object {},
  "language": "en",
  "pagination": "11/20",
  "section": "ALL",
}
`;

exports[`QueryContext section can be changed through the reducer 1`] = `
Object {
  "filters": Object {},
  "language": "en",
  "pagination": "0/10",
  "section": "LATEST",
}
`;
